metaflow/plugins/kfp/kfp.py

```diff
@@ -28,35 +29,166 @@ def get_ordered_steps(graph):
+def step_op_func(step_name: str,
+                 code_url: str,
+                 ds_root: str,
+                 run_id: str,
+                 task_id: str,
+                 prev_step_name: str,
+                 prev_task_id: str) -> NamedTuple('StepOutput', [('ds_root', str),
+                                                         ('run_id', str),
+                                                         ('next_step', str),
+                                                         ('next_task_id', str),
+                                                         ('current_step', str),
+                                                         ('current_task_id', str)]
+                                                      ):
     """
-    Method to create a kfp container op that corresponds to a single step in the flow.
+    Function used to create a KFP container op (see: `step_container_op`) that corresponds to a single step in the flow.
 
-    TODO: This does not maintain state. The custom pre-start command used below would be removed once we have state accessible across KFP steps.
-    TODO: The public docker is a copy of the internal docker image we were using (borrowed from aip-kfp-example). Check if any stage here may need to be further modified later.
     """
+    import subprocess
+    from collections import namedtuple
+
+    print("\n----------RUNNING: CODE DOWNLOAD from URL---------")
+    subprocess.call(["curl -o helloworld.py {}".format(code_url)], shell=True)
+
+    print("\n----------RUNNING: KFP Installation---------------")
+    subprocess.call(["pip3 install kfp"], shell=True) # TODO: Remove this once KFP is added to dependencies
+
+    print("\n----------RUNNING: METAFLOW INSTALLATION----------")
+    subprocess.call(["pip3 install --user --upgrade git+https://github.com/zillow/metaflow.git@state-integ-s3"],
+                    shell=True)
+
+    print("\n----------RUNNING: MAIN STEP COMMAND--------------")
+    S3_BUCKET = "s3://workspace-zillow-analytics-stage/aip/metaflow"
+    S3_AWS_ARN = "arn:aws:iam::170606514770:role/dev-zestimate-role"
+
+    define_s3_env_vars = 'export METAFLOW_DATASTORE_SYSROOT_S3="{}" && export METAFLOW_AWS_ARN="{}" '.format(S3_BUCKET,
+                                                                                                             S3_AWS_ARN)
+    define_username = 'export USERNAME="kfp-user"'
+    python_cmd = "python helloworld.py --datastore s3 --datastore-root {0} step {1} --run-id {2} " \
+                 "--task-id {3} --input-paths {2}/{4}/{5}".format(ds_root, step_name, run_id,
+                                                                  task_id, prev_step_name, prev_task_id)
+    final_run_cmd = f'{define_username} && {define_s3_env_vars} && {python_cmd}'
+
+    print("RUNNING COMMAND: ", final_run_cmd)
+    proc = subprocess.run(final_run_cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
```

Comment:
If the actual step code itself fails for some reason looks like we've taken care of piping the logs from `stderr` (or there is a TODO to fix it which I'm guessing will be resolved in this PR). How are we propagating the exit-code to the Metaflow scheduler, is this something that is handled by this code at the moment?

Reply:
This is not handled actually and I'd noticed the problem while testing and have it documented here in [AIP-1429](https://zbrt.atl.zillow.net/browse/AIP-1429)